types/core/index.d.ts

```diff
@@ -263,7 +263,7 @@ export interface Chart<
   buildOrUpdateScales(): void;
   buildOrUpdateControllers(): void;
   reset(): void;
-  update(mode?: string): void;
+  update(mode?: UpdateMode): void;
```

Comment:
This can be a custom mode too, so `UpdateMode | string`?

Reply:
If it can be custom, then yes, you are right. Note, however, that this type is used in several places. Could be a custom mod there too? (in `DatasetController`)

Reply:
Yes, its the same parameter passed along.
https://www.chartjs.org/docs/master/developers/api#updatemode

Reply:
Complicated... On the one hand, this is correct, on the other, the meaning in literal types is lost.
I mean that `(<literal type of string> | string) === string`.

I think that even if this is correct, it is better not to specify the type for the custom mode. And indicate in the documentation that casting must be done for the custom mode. Because this is an advanced development.
```ts
chartInstance.update('custom' as UpdateMode);
```
What do you think?

Also, most likely, this problem is relevant for the types of charts. Even though intellij provides autocomplete, but the VS Code does not. I think it's worth removing the string type from the chart types as well...

Reply:
Sounds good. I don't have the knowledge to form an opinion though, so will have to leave this with others. @etimberg @sgratzl

Reply:
This is at the edge of my experience with typescript. I think updating the docs and recommending a cast would be fine. I don't know if there are any custom modes yet since this is a pretty new feature.

Reply:
After some thought and enumeration of different options (generic type, casting, etc), it seems to me that this option will be the best. But let me clarify why I think so. TS helps to understand when the wrong attribute was used, it is especially useful for typos. If we add a custom mode, the TS will think that the typo is a custom mode. It will be difficult to understand the issue. Advanced developers will have to cast (cheat the TS), and they will do it on purpose.
Following the same idea, I removed the custom type for the chart types.

Reply:
The right way to do this kind of things in typescript is usually to use `enum`, because `enum`, unlike type alias, can be augmented using declaration merging. This means that we can do this: [example](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBEDFACSXkQ8G4USUVMUqPYiFAAGEHGRUAdQABvUAvakAOQAoqAALygADk0AykFl-AAQhUntK5dJHLLQABfYLJLiwVAUeBediCSDcxx8loAClQLXyXJ5dqIAEpBYoSiUEJxBPB3AFoPBmI6Wh7FAaiipaaBLT90dxmcBgmyOTbefyhZyHsQcAB5ACymtlsEt6EQupjiaz7sg9vrBwCIolXpUhYA0oo626Dk3+y0AnmCyWO2Bu4ogA), which preserves the entire type safety (without the need to cast, which is not safe by nature) both in chart.js library and plugin or user code.

However typescript enums are also different from type aliases because they generate actual javascript code corresponding to the enum object (so that enum can be used as a plain object in javascript code). If we cannot afford that there is also `const enum`, but const enums can have some issues when exported in ES6 modules, and are generally preferable in private code...

Reply:
Right. But then users have to use enum for basic use too. They cannot just write "line" value in the chart type, it will be a big move...

Reply:
I still think DX is more important when the correct type. I don't think casting for advanced users is a big issue.
But in the end, the final decision will be @kurkle @etimberg.

@emmcbd  what do you think?

Reply:
@xr0master It also possible to use "line" values with enum types, see [here](https://www.typescriptlang.org/play?#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBEDFACSXkQ8G4USUVMUqPYiFAAGEHGRUAdQABvNwZFAAXlAAHJoCKJfxpI5QOKJfKyBLQABfYLJLiwVAUeBediCSDcxx8loAClQLXyoHikEMqWtRBJpt5BwAlILFCUSghOIJ4O4AtB4MwrS0PYpNUUVLTQEafujuMzgME2Ry3eaiIKTEb0BylbB80g1THFImswcLVKZV6VAB5ADSFeNVctEuLggLEvrYGbrZNPOzkBrXYLACZe6AVE2G4ogA)

Reply:
If we can use the enum with custom types, that would work for me.

Reply:
@emmcbd let's use a more real-world scenario than one code per file:
https://codesandbox.io/s/nifty-sid-j0mub
Can you show how to extend enum? Thank you.

btw with enum, inside of interface.d, TS doesn't care if the wrong key is used:
![image](https://user-images.githubusercontent.com/8962340/92486903-49766980-f1f5-11ea-9cb6-8c665f13b51a.png)

Reply:
@xr0master here you go
https://codesandbox.io/s/broken-sun-crj4n

In a real project using chart.js types, you would do `declare module 'chart.js'` in a .d.ts file at the root, so the augmented enum would becom available in the whole project.

Reply:
Great! Thank. I've learned something new about enums.
Pushed changes about enums

Reply:
You're welcome! Now I think you can also remove the TYPE generic everywhere, and directly use IChartType type (if I am right, the generic was only needed to pass a custom type)